File: src/shmif/arcan_shmif_filehelper.c
#include <arcan_shmif.h>
#include <pthread.h>
#include "platform/shmif_platform.h"
#include <errno.h>
#include <sys/stat.h>

static bool write_buffer(int fd, char* inbuf, size_t inbuf_sz)
{
	while(inbuf_sz){
		ssize_t nr = write(fd, inbuf, inbuf_sz);
		if (-1 == nr){
			if (errno == EAGAIN || errno == EINTR)
				continue;
			return false;
		}
		inbuf += nr;
		inbuf_sz -= nr;
	}
	return true;
}

static void* copy_thread(void* inarg)
{
	int* fds = inarg;
	char inbuf[4096];
	int8_t sc = 0;

	size_t tot = 0;
	size_t acc = 0;
	size_t last_acc = 0;
	uint64_t time_last = arcan_timemillis();
	static const size_t report_mb = 10;

	struct stat fs;

/* might not remain accurate but fair to keep around, slightly more
 * diligent is re-stating on boundary invalidation (st_size < acc) */
	if (-1 != fstat(fds[0], &fs) && fs.st_size > 0){
		tot = fs.st_size;
	}

/* depending on type and OS, there are a number of options e.g. sendfile,
 * splice, sosplice, ... right now just use a slow/safe */
	for(;;){
		ssize_t nr = read(fds[0], inbuf, sizeof(inbuf));
		if (-1 == nr){
			if (errno == EAGAIN || errno == EINTR)
				continue;
			sc = -1;
			break;
		}
		if (0 == nr){
			break;
		}
		else if (!write_buffer(fds[1], inbuf, nr)){
			sc = -2;
			break;
		}

/* PIPE_BUF is required to be >= 512 on POSIX, only update every n megabytes or
 * every second or so as to not block unnecessarily on reporting while still
 * being responsive. */
		if (fds[3] & SHMIF_BGCOPY_PROGRESS){
			acc += nr;

			if (acc - last_acc > report_mb * 1024 * 1024 ||
				arcan_timemillis() - time_last > 1000){

				time_last = arcan_timemillis();
				int n = snprintf(inbuf,
					sizeof(inbuf), "%zu:%zu:%zu\n", (size_t) nr, acc, tot);
				write(fds[2], inbuf, n);
			}
		}
	}

	if (!(fds[3] & SHMIF_BGCOPY_KEEPIN))
		close(fds[0]);
	if (!(fds[3] & SHMIF_BGCOPY_KEEPOUT))
		close(fds[1]);

	if (-1 != fds[2]){
		if (fds[3] & SHMIF_BGCOPY_PROGRESS){
			int n = snprintf(inbuf, sizeof(inbuf), "%d:%zu:%zu\n", sc, acc, tot);
			while (-1 == write(fds[2], inbuf, n) &&
				(errno == EAGAIN || errno == EINTR)){}
		}
		else
			while (-1 == write(fds[2], &sc, 1) &&
				(errno == EAGAIN || errno == EINTR)){}

		close(fds[2]);
	}

	free(fds);
	return NULL;
}

void arcan_shmif_bgcopy(
	struct arcan_shmif_cont* c, int fdin, int fdout, int sigfd, int fl)
{
	int* fds = malloc(sizeof(int) * 4);
	if (!fds)
		return;

	fds[0] = fdin;
	fds[1] = fdout;
	fds[2] = sigfd;
	fds[3] = fl;

/* options, fork or thread */
	pthread_t pth;
	pthread_attr_t pthattr;
	pthread_attr_init(&pthattr);
	pthread_attr_setdetachstate(&pthattr, PTHREAD_CREATE_DETACHED);

	if (-1 == pthread_create(&pth, &pthattr, copy_thread, fds)){
		if (!(fl & SHMIF_BGCOPY_KEEPIN))
			close(fdin);
		if (!(fl & SHMIF_BGCOPY_KEEPOUT))
			close(fdout);
		if (-1 != sigfd){
			int8_t ch = -3;
			write(sigfd, &ch, 1);
		}
		free(fds);
	}
}

/*
 * Missing: special behavior for SHMIF_RHINT_SUBREGION_CHAIN, setup chain of
 * atomic [uint32_t, bitfl] and walk from first position to last free. Bitfl
 * marks both the server side "I know this" and client side "synch this or
 * wait".
 *
 * If there's not enough store left to walk on, signal video and wait (unless
 * noblock flag is set)
 */

#ifdef __LINUX
char* arcan_shmif_bchunk_resolve(
	struct arcan_shmif_cont* C, struct arcan_event* bev)
{
	char buf[24];
	if (bev->category != EVENT_TARGET ||
		(bev->tgt.kind != TARGET_COMMAND_BCHUNK_IN &&
		 bev->tgt.kind != TARGET_COMMAND_BCHUNK_OUT))
		return NULL;

	char* mbuf = malloc(PATH_MAX);
	if (!mbuf)
		return mbuf;

	snprintf(buf, sizeof(buf), "/proc/self/fd/%d", bev->tgt.ioevs[0].iv);
	ssize_t rv = readlink(buf, mbuf, PATH_MAX);

	if (-1 == rv || mbuf[0] != '/')
		return NULL;

	struct stat base, comp;
	if (-1 == fstat(bev->tgt.ioevs[0].iv, &base) ||
			-1 == stat(mbuf, &comp) || base.st_ino != comp.st_ino){
		free(mbuf);
		return NULL;
	}

	return mbuf;
}
/* OpenBSD has no solution, FreeBSD / OSX has a fcntl that can be used */
#else
char* arcan_shmif_bchunk_resolve(
	struct arcan_shmif_cont* C, struct arcan_event* bev)
{
	return NULL;
}
#endif

File: src/shmif/arcan_shmif_filehelper.zig
// Zig reimplementation of arcan_shmif_filehelper.c
// Drop-in C-ABI-compatible replacement for filehelper functions.
//
// Exports: arcan_shmif_bgcopy, arcan_shmif_bchunk_resolve
//
const std = @import("std");
const builtin = @import("builtin");
const is_freestanding = (builtin.os.tag == .freestanding);
const c = @import("shmif_types");

// Extern C declarations

extern fn malloc(size: usize) ?*anyopaque;
extern fn free(ptr: ?*anyopaque) void;
extern fn close(fd: c_int) c_int;
extern fn read(fd: c_int, buf: ?*anyopaque, count: usize) isize;
extern fn write(fd: c_int, buf: ?*const anyopaque, count: usize) isize;
extern fn snprintf(noalias buf: [*c]u8, size: usize, noalias fmt: [*c]const u8, ...) c_int;
extern fn arcan_timemillis() u64;
extern fn readlink(
    path: [*c]const u8,
    buf: [*c]u8,
    bufsiz: usize,
) isize;

const stat_t = c.struct_stat;
extern fn fstat(fd: c_int, buf: *stat_t) c_int;
extern fn stat(path: [*c]const u8, buf: *stat_t) c_int;

extern fn pthread_create(
    thread: *std.c.pthread_t,
    attr: ?*const std.c.pthread_attr_t,
    start_routine: *const fn (?*anyopaque) callconv(.c) ?*anyopaque,
    arg: ?*anyopaque,
) c_int;
extern fn pthread_attr_init(attr: *std.c.pthread_attr_t) c_int;
extern fn pthread_attr_setdetachstate(attr: *std.c.pthread_attr_t, detachstate: c_int) c_int;

const PTHREAD_CREATE_DETACHED: c_int = 1;
const PATH_MAX: usize = 4096;

// static helper: write_buffer

fn write_buffer(fd: c_int, inbuf_in: [*c]u8, inbuf_sz_in: usize) bool {
    var inbuf: [*c]u8 = inbuf_in;
    var inbuf_sz = inbuf_sz_in;
    while (inbuf_sz > 0) {
        const nr = write(fd, @as(?*const anyopaque, @ptrCast(inbuf)), inbuf_sz);
        if (nr == -1) {
            const err = std.c._errno().*;
            if (err == c.EAGAIN or err == c.EINTR)
                continue;
            return false;
        }
        inbuf += @intCast(nr);
        inbuf_sz -= @intCast(nr);
    }
    return true;
}

// static helper: copy_thread

fn copy_thread(inarg: ?*anyopaque) callconv(.c) ?*anyopaque {
    const fds: *[4]c_int = @ptrCast(@alignCast(inarg));
    var inbuf: [4096]u8 = undefined;
    var sc: i8 = 0;

    var acc: usize = 0;
    var last_acc: usize = 0;
    const report_mb: usize = 10;
    var time_last: u64 = arcan_timemillis();

    var tot: usize = 0;
    var fs: stat_t = std.mem.zeroes(stat_t);

    // might not remain accurate but fair to keep around
    if (fstat(fds[0], &fs) != -1 and fs.st_size > 0) {
        tot = @intCast(fs.st_size);
    }

    // depending on type and OS, there are a number of options e.g. sendfile,
    // splice, sosplice, ... right now just use a slow/safe
    while (true) {
        const nr = read(fds[0], @as(?*anyopaque, @ptrCast(&inbuf)), inbuf.len);
        if (nr == -1) {
            const err = std.c._errno().*;
            if (err == c.EAGAIN or err == c.EINTR)
                continue;
            sc = -1;
            break;
        }
        if (nr == 0) {
            break;
        } else if (!write_buffer(fds[1], @as([*c]u8, @ptrCast(&inbuf)), @intCast(nr))) {
            sc = -2;
            break;
        }

        // PIPE_BUF is required to be >= 512 on POSIX, only update every n megabytes or
        // every second or so as to not block unnecessarily on reporting
        if (fds[3] & @as(c_int, c.SHMIF_BGCOPY_PROGRESS) != 0) {
            acc += @intCast(nr);

            if (acc - last_acc > report_mb * 1024 * 1024 or
                arcan_timemillis() - time_last > 1000)
            {
                last_acc = acc;
                time_last = arcan_timemillis();
                const n = snprintf(
                    @as([*c]u8, @ptrCast(&inbuf)),
                    inbuf.len,
                    "%zu:%zu:%zu\n",
                    @as(usize, @intCast(nr)),
                    acc,
                    tot,
                );
                _ = write(fds[2], @as(?*const anyopaque, @ptrCast(&inbuf)), @intCast(n));
            }
        }
    }

    if (fds[3] & @as(c_int, c.SHMIF_BGCOPY_KEEPIN) == 0)
        _ = close(fds[0]);
    if (fds[3] & @as(c_int, c.SHMIF_BGCOPY_KEEPOUT) == 0)
        _ = close(fds[1]);

    if (fds[2] != -1) {
        if (fds[3] & @as(c_int, c.SHMIF_BGCOPY_PROGRESS) != 0) {
            const n = snprintf(
                @as([*c]u8, @ptrCast(&inbuf)),
                inbuf.len,
                "%d:%zu:%zu\n",
                @as(c_int, sc),
                acc,
                tot,
            );
            while (write(fds[2], @as(?*const anyopaque, @ptrCast(&inbuf)), @intCast(n)) == -1) {
                const err = std.c._errno().*;
                if (err != c.EAGAIN and err != c.EINTR)
                    break;
            }
        } else {
            while (write(fds[2], @as(?*const anyopaque, @ptrCast(&sc)), 1) == -1) {
                const err = std.c._errno().*;
                if (err != c.EAGAIN and err != c.EINTR)
                    break;
            }
        }

        _ = close(fds[2]);
    }

    free(@as(?*anyopaque, @ptrCast(fds)));
    return null;
}

// arcan_shmif_bgcopy

export fn arcan_shmif_bgcopy(
    _: ?*c.struct_arcan_shmif_cont,
    fdin: c_int,
    fdout: c_int,
    sigfd: c_int,
    fl: c_int,
) void {
    if (is_freestanding) return;
    const raw = malloc(@sizeOf(c_int) * 4) orelse return;
    const fds: *[4]c_int = @ptrCast(@alignCast(raw));
    fds[0] = fdin;
    fds[1] = fdout;
    fds[2] = sigfd;
    fds[3] = fl;

    // options, fork or thread
    var pth: std.c.pthread_t = undefined;
    var pthattr: std.c.pthread_attr_t = undefined;
    _ = pthread_attr_init(&pthattr);
    _ = pthread_attr_setdetachstate(&pthattr, PTHREAD_CREATE_DETACHED);

    if (pthread_create(&pth, &pthattr, &copy_thread, @as(?*anyopaque, @ptrCast(fds))) != 0) {
        if (fl & @as(c_int, c.SHMIF_BGCOPY_KEEPIN) == 0)
            _ = close(fdin);
        if (fl & @as(c_int, c.SHMIF_BGCOPY_KEEPOUT) == 0)
            _ = close(fdout);
        if (sigfd != -1) {
            var ch: i8 = -3;
            _ = write(sigfd, @as(?*const anyopaque, @ptrCast(&ch)), 1);
        }
        free(raw);
    }
}

// arcan_shmif_bchunk_resolve

export fn arcan_shmif_bchunk_resolve(
    _: ?*c.struct_arcan_shmif_cont,
    bev: ?*c.arcan_event,
) [*c]u8 {
    if (is_freestanding) return null;
    if (comptime builtin.os.tag == .linux) {
        const ev = bev orelse return null;
        if (ev.category().* != c.EVENT_TARGET or
            (ev.tgt().kind != c.TARGET_COMMAND_BCHUNK_IN and
                ev.tgt().kind != c.TARGET_COMMAND_BCHUNK_OUT))
            return null;

        var buf: [24]u8 = undefined;
        _ = snprintf(
            @as([*c]u8, @ptrCast(&buf)),
            buf.len,
            "/proc/self/fd/%d",
            ev.tgt().ioevs[0].iv,
        );

        const mbuf: [*c]u8 = @ptrCast(malloc(PATH_MAX) orelse return null);

        const rv = readlink(
            @as([*c]const u8, @ptrCast(&buf)),
            mbuf,
            PATH_MAX,
        );

        if (rv == -1 or mbuf[0] != '/') {
            free(@as(?*anyopaque, @ptrCast(mbuf)));
            return null;
        }

        var base_stat: stat_t = std.mem.zeroes(stat_t);
        var comp_stat: stat_t = std.mem.zeroes(stat_t);
        if (fstat(ev.tgt().ioevs[0].iv, &base_stat) == -1 or
            stat(mbuf, &comp_stat) == -1 or
            base_stat.st_ino != comp_stat.st_ino)
        {
            free(@as(?*anyopaque, @ptrCast(mbuf)));
            return null;
        }

        return mbuf;
    } else {
        // OpenBSD has no solution, FreeBSD / OSX has a fcntl that can be used
        return null;
    }
}
